contrib/epee/include/serialization/wire/write.h

```diff
@@ -0,0 +1,290 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <boost/utility/string_ref.hpp>
+#include <boost/range/size.hpp>
+#include <cstdint>
+#include <system_error>
+#include <type_traits>
+
+#include "byte_slice.h"
+#include "serialization/wire/error.h"
+#include "serialization/wire/field.h"
+#include "serialization/wire/traits.h"
+#include "span.h"
+
+/*
+  Custom types (e.g type `type` in namespace `ns`) can define an output function by:
+    * `namespace wire { template<> struct is_array<ns::type> : std::true_type {}; }`
+    * `namespace wire { template<> struct is_blob<ns::type> : std::true_type {}; }`
+    * `namespace wire { void write_bytes(writer&, const ns::type&); }`
+    * `namespace ns { void write_bytes(wire::writer&, const type&); }`
+
+  See `wrappers.h` for `is_array` requirements, and `traits.h` for `is_blob`
+  requirements. `write_bytes` function can also specify derived type for faster
+  output (i.e.  `namespace ns { void write_bytes(wire::epee_writer&, type&); }`).
+  Using the derived type allows the compiler to de-virtualize and allows for
+  custom functions not defined by base interface. Using base interface allows
+  for multiple formats with minimal instruction count. */
+
+namespace wire
+{
+  //! Interface for converting C/C++ objects to "wire" (byte) formats.
+  struct writer
+  {
+    writer() = default;
+
+    virtual ~writer() noexcept;
+
+    //! By default, insist on retrieving array size before writing array
+    static constexpr std::true_type need_array_size() noexcept { return{}; }
+
+    virtual void boolean(bool) = 0;
+
+    virtual void integer(int) = 0;
+    virtual void integer(std::intmax_t) = 0;
+
+    virtual void unsigned_integer(unsigned) = 0;
+    virtual void unsigned_integer(std::uintmax_t) = 0;
```

Comment:
Is there any guarantee that `int` and `std::intmax_t` are different types on every system? Also, I haven't run benchmarks, but would it be a worthy tradeoff to crunch each `integer` and `unsigned_integer` virtual methods into one each? In other words, is the expanded vtables / extra binary size / worse caching worth not having to cast the integers?

Reply:
The only advantage is 32-bit systems. Theoretically some platform could use 128-bit integers on a 64-bit cpu (making the overload useful again), but that doesn't seem likely. It might be worth dropping to simplify things a bit more.

And I don't think its possible for `int` and `std::intmax_t` to have an issue, because both `long` and `long long` are required by the standard, and `std::intmax_t` must be the largest possible type on the platform. Even if they are all 32-bit, they should be unique types (afaik).

Of course, if I drop the overload entirely then it doesn't matter.

Reply:
I think from a semantics point of view, since there is no difference between these two methods anyways that the overloads should be taken away IMO; it simplifies the interface. If there was a difference (see `serde`), then it would make sense to keep them in

Reply:
I removed the `int` and `unsigned` overloads, only the `intmax_t` and `uintmax_t` remain.

Reply:
The unit tests just reminded me of one issue with removing the `unsigned` and `int` - when writing "through the base class" epee will _always_ write 64-bit integers. This can be prevented by checking the value and compressing the output (like my message pack implementation); the split previously allowed for a quick way to compress the types.

I think I'm going to make a change where the epee class always compresses the integer to the smallest possible (like message pack does now).

Reply:
Nevermind, I remember why I cannot. MessagePack has a tag per array item, allowing you to compress individual integers, but epee does not. So if the base-class is used for writing epee data, then it will _always_ use 64-bit integers for outputs.

Reply:
Also some binary formats (like bincode) have fixed size integer types. We might want to, in the future, add specific methods for this. Think like Rust's serde's `Serializer` interface with `serialize_u8`, `serialize_i16`, `serialize_u32`, `serialize_u64`, etc. By default, for most formats, these methods could feed directly into `integer()` and `unsigned_integer()`. But I guess we shouldn't over-engineer from the outset.

Reply:
I'm not sure what you mean here - but MessagePack and Epee both have fixed integer types. Neither has variable size actually (whereas Protobuf does). That's not the issue. The problem is _every integer comes with its own tag_ for size with MessagePack, whereas with Epee there is one-fixed tag for every integer in an array.

I already wrote the Epee writer to serialize the exact fixed size when using the derived class (like serde), but it now fallsback to uint64 when using the base class. You can overload the write function multiple times if its a problem, so I don't see a need to engineer this further.

The MessagePack implement just compresses the value per integer, which is what most of the MessagePack serializers do.

Reply:
What I mean is that in both of those cases, Epee and MessagePack, those formats are *self-describing*. You can determine the size of integers at runtime. Because of that property, you can be a little more willy-nilly casting integers to and forth into the format, and it isn't actually that important that we provide methods for every fixed size of integer. By contrast, in bincode, for example, this serialization information is assumed to be known beforehand, which saves space and computation, but if you don't know what size to deserialize an integer as, then reading from that point on might just give you garbage.

In MessagePack, since this tag is per-element (I'm assuming what you've told me I don't actually know anything about the Messagepack format), then you don't have to actually call a special `integer` method because the `writer` in this case could just assume you always want to compress if the value is small enough. And, likewise, since the `reader` can see the tags before deserializing the integers, we only need 2 methods for integers: signed and unsigned, and just cast everything accordingly.

I think it's fine like it is now with just `integer()/unsigned_integer()`, but if we wanted to support formats which use multiple sizes of integers but don't encode that integer size information in the data, then we would need more expressive interface with methods for each fixed integer size.

Reply:
Yes, this is why I replaced everything but the internal crypto note serialization format.